src/util/domains.py

```diff
@@ -45,7 +45,9 @@
 
 # import from domains/expression
 try:
-    import odoo.domains as _dom
+    if not version_gte("saas~18.2"):
```

Comment:
Shouldn't we make this 18.1 as the [original](https://github.com/odoo/upgrade-util/pull/99) patch (and the odoo/odoo part) was merged there?

Reply:
I waited especially for changes in domains for **18.2**.
The main one is having semantically equivalent complements of domains. For example: `['!', ('opt_date', '>', '2020-01-01')]` is not equivalent to `[('opt_date', '<=', '2020-01-01')]` because of null value handling - so when building/translating domains this is now taken into account and the rewrite does not happen. Before 18.2, `distribute_not` has this "bug" and it is part of domain building.

Reply:
Don't we need to keep then both? Or we don't care about upgrading to 18.1 with the pre-18.1 way?

Reply:
In 18.1, you are migrated using the old `normalize_domain` function. As of 18.2, you are migrated using the Domain object which distributes the not values. In 18.2, we fix the distribution of *not* for some edge cases, so upgrading using either method is correct.